lib/beamchmark/suite.ex

```diff
@@ -0,0 +1,95 @@
+defmodule Beamchmark.Suite do
+  @moduledoc """
+  The module defines a struct representing a single run of benchmark.
+  It is responsible for benchmarking and saving/loading the results.
+  """
+
+  alias Beamchmark.Scenario
+  alias __MODULE__.{Configuration, SystemInfo, Measurements}
+
+  @type t :: %__MODULE__{
+          scenario: Scenario.t(),
+          configuration: Configuration.t(),
+          system_info: SystemInfo.t(),
+          measurements: Measurements.t() | nil
+        }
+
+  @enforce_keys [
+    :scenario,
+    :configuration,
+    :system_info,
+    :measurements
+  ]
+  defstruct @enforce_keys
+
+  @suite_filename "suite"
+  @old_suite_filename "suite_old"
+
+  @spec init(Scenario.t(), Configuration.t()) :: t()
+  def init(scenario, %Configuration{} = configuration) do
+    implements_scenario? =
+      scenario.module_info(:attributes)
+      |> Keyword.get(:behaviour, [])
+      |> Enum.member?(Scenario)
+
+    unless implements_scenario? do
+      raise "#{inspect(scenario)} is not a module implementing #{inspect(Scenario)} behaviour."
+    end
+
+    %__MODULE__{
+      scenario: scenario,
+      configuration: configuration,
+      system_info: SystemInfo.init(),
+      measurements: nil
+    }
+  end
+
+  @spec run(t()) :: t()
+  def run(%__MODULE__{scenario: scenario, configuration: config} = suite) do
+    # TODO: what about scenarios that take less than `delay + duration` seconds or run indefinitely?
```

Comment:
> what about scenarios that take less than `delay + duration`

I think that's not our problem

> or run indefinitely?

Maybe we should add one more function to our scenario e.g. `shutdown` that will be called after gathering stats?

Reply:
Given that we want to benchmark long-running applications, it seems reasonable 👍